syntax = "proto3";
package auth_service;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

option go_package = "github.com/milkyhoop/auth-service/internal/delivery/pb/auth;auth";

// AuthService - Enterprise Authentication & User Management
// Features: tenant isolation, token rotation, session management, audit compliance
service AuthService {
  // Core Authentication
  rpc Register (RegisterRequest) returns (RegisterResponse);
  rpc Login (LoginRequest) returns (LoginResponse);
  rpc RefreshToken (RefreshTokenRequest) returns (RefreshTokenResponse);
  rpc ValidateToken (ValidateTokenRequest) returns (ValidateTokenResponse);
  rpc RevokeToken (RevokeTokenRequest) returns (RevokeTokenResponse);
  rpc GetUserProfile (UserProfileRequest) returns (UserProfileResponse);
  
  // Token Management & Security
  rpc IntrospectToken (IntrospectTokenRequest) returns (IntrospectTokenResponse);
  rpc ListActiveSessions (ListActiveSessionsRequest) returns (ListActiveSessionsResponse);
  rpc RevokeSession (RevokeSessionRequest) returns (RevokeSessionResponse);
  rpc RevokeAllSessions (RevokeAllSessionsRequest) returns (RevokeAllSessionsResponse);
  
  // Audit & Compliance
  rpc LogAuditEvent (AuditEventRequest) returns (AuditEventResponse);
  rpc GetAuditTrail (AuditTrailRequest) returns (AuditTrailResponse);
  
  // Service Operations
  rpc GetServiceInfo (google.protobuf.Empty) returns (ServiceInfoResponse);
  rpc HealthCheck (google.protobuf.Empty) returns (google.protobuf.Empty);
}

// ========================================
// CORE AUTHENTICATION MESSAGES
// ========================================

message RegisterRequest {
  string email = 1;
  string password = 2;
  string name = 3;
  string username = 4;
  string tenant_id = 5;
  string device_info = 6;
  map<string, string> metadata = 7; // ip_address, user_agent, location
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
  string user_id = 3;
  string tenant_id = 4;
  string access_token = 5;
  string refresh_token = 6;
  string session_id = 7;
  int64 expires_in = 8; // seconds until access token expires
  string role = 9;
  google.protobuf.Timestamp issued_at = 10;
}

message LoginRequest {
  string email = 1;
  string password = 2;
  string device_info = 3;
  string session_id = 4; // optional, for session continuation
  map<string, string> metadata = 5;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
  string user_id = 3;
  string tenant_id = 4;
  string access_token = 5;
  string refresh_token = 6;
  string session_id = 7;
  int64 expires_in = 8;
  string role = 9;
  google.protobuf.Timestamp issued_at = 10;
  google.protobuf.Timestamp last_login = 11;
  bool requires_password_change = 12;
}

// ========================================
// TOKEN ROTATION (ENTERPRISE HARDENED)
// ========================================

message RefreshTokenRequest {
  string refresh_token = 1;
  string device_info = 2;
  string session_id = 3;
  map<string, string> metadata = 4;
}

message RefreshTokenResponse {
  bool success = 1;
  string message = 2;
  string access_token = 3;
  string refresh_token = 4; // NEW rotated token
  string tenant_id = 5;
  string user_id = 6;
  string role = 7;
  string previous_token_id = 8; // for audit trail
  string session_id = 9;
  int64 expires_in = 10;
  bool token_rotated = 11; // indicates token was rotated
  google.protobuf.Timestamp issued_at = 12;
}

message ValidateTokenRequest {
  string access_token = 1;
  string required_role = 2; // optional role check
  string required_tenant = 3; // optional tenant verification
  map<string, string> metadata = 4;
}

message ValidateTokenResponse {
  bool valid = 1;
  string user_id = 2;
  string tenant_id = 3;
  string role = 4;
  string session_id = 5;
  string message = 6;
  int64 expires_in = 7; // seconds until expiration
  bool requires_refresh = 8; // token near expiry
  repeated string permissions = 9; // user permissions
}

message RevokeTokenRequest {
  string refresh_token = 1;
  string revocation_reason = 2; // LOGOUT, SECURITY_BREACH, ADMIN_ACTION
  string session_id = 3;
  map<string, string> metadata = 4;
}

message RevokeTokenResponse {
  bool success = 1;
  string message = 2;
  int32 revoked_sessions = 3;
  string audit_id = 4; // for tracking
}

// ========================================
// TOKEN INTROSPECTION
// ========================================

message IntrospectTokenRequest {
  string token = 1; // access or refresh token
  string requesting_service = 2; // for audit
}

message IntrospectTokenResponse {
  bool active = 1;
  string user_id = 2;
  string tenant_id = 3;
  string role = 4;
  string session_id = 5;
  string token_type = 6; // ACCESS or REFRESH
  int64 exp = 7; // expiration timestamp
  int64 iat = 8; // issued at timestamp
  string jti = 9; // JWT ID
  repeated string permissions = 10;
  map<string, string> claims = 11; // additional token claims
}

// ========================================
// SESSION MANAGEMENT
// ========================================

message ListActiveSessionsRequest {
  string user_id = 1;
  string requesting_user_id = 2; // for audit
  map<string, string> metadata = 3;
}

message SessionInfo {
  string session_id = 1;
  string device_info = 2;
  string ip_address = 3;
  string user_agent = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp last_activity = 6;
  bool is_current = 7; // current requesting session
  bool is_active = 8;
  string location = 9; // geo location if available
}

message ListActiveSessionsResponse {
  bool success = 1;
  repeated SessionInfo sessions = 2;
  int32 total_sessions = 3;
  string message = 4;
}

message RevokeSessionRequest {
  string session_id = 1;
  string requesting_user_id = 2;
  string revocation_reason = 3;
  map<string, string> metadata = 4;
}

message RevokeSessionResponse {
  bool success = 1;
  string message = 2;
  string audit_id = 3;
}

message RevokeAllSessionsRequest {
  string user_id = 1;
  bool exclude_current = 2; // don't revoke current session
  string revocation_reason = 3;
  map<string, string> metadata = 4;
}

message RevokeAllSessionsResponse {
  bool success = 1;
  string message = 2;
  int32 revoked_sessions = 3;
  string audit_id = 4;
}

// ========================================
// USER PROFILE
// ========================================

message UserProfileRequest {
  string user_id = 1;
  string requesting_user_id = 2;
  map<string, string> metadata = 3;
}

message UserProfileResponse {
  bool success = 1;
  string user_id = 2;
  string tenant_id = 3;
  string email = 4;
  string name = 5;
  string username = 6;
  string role = 7;
  bool is_active = 8;
  bool email_verified = 9;
  google.protobuf.Timestamp created_at = 10;
  google.protobuf.Timestamp last_login = 11;
  repeated string permissions = 12;
  map<string, string> profile_data = 13;
}

// ========================================
// AUDIT & COMPLIANCE
// ========================================

message AuditEventRequest {
  string user_id = 1;
  string tenant_id = 2;
  string session_id = 3;
  string action = 4; // LOGIN, LOGOUT, TOKEN_REFRESH, CROSS_TENANT_ATTEMPT
  string resource_type = 5; // USER, TOKEN, SESSION, PROFILE
  string resource_id = 6;
  string violation_type = 7; // CROSS_TENANT_ACCESS, SUSPICIOUS_ACTIVITY
  map<string, string> details = 8;
  string ip_address = 9;
  string user_agent = 10;
  google.protobuf.Timestamp timestamp = 11;
}

message AuditEventResponse {
  bool success = 1;
  string audit_id = 2;
  string message = 3;
}

message AuditTrailRequest {
  string user_id = 1;
  string tenant_id = 2;
  string action_filter = 3;
  google.protobuf.Timestamp start_time = 4;
  google.protobuf.Timestamp end_time = 5;
  int32 limit = 6;
  string cursor = 7; // for pagination
}

message AuditEntry {
  string audit_id = 1;
  string user_id = 2;
  string tenant_id = 3;
  string action = 4;
  string resource_type = 5;
  string violation_type = 6;
  string ip_address = 7;
  google.protobuf.Timestamp timestamp = 8;
  map<string, string> details = 9;
}

message AuditTrailResponse {
  bool success = 1;
  repeated AuditEntry entries = 2;
  string next_cursor = 3;
  int32 total_count = 4;
}

// ========================================
// SERVICE OBSERVABILITY
// ========================================

message ServiceInfoResponse {
  string service_name = 1;
  string version = 2;
  string build_time = 3;
  string environment = 4;
  bool tenant_isolation_enabled = 5;
  bool audit_logging_enabled = 6;
  bool token_rotation_enabled = 7;
  bool session_management_enabled = 8;
  map<string, string> security_features = 9;
  map<string, string> capabilities = 10;
}